protocol/src/main/kotlin/models/Message.kt

```diff
@@ -1,89 +1,83 @@
 package models
 
-import Mapper
 import com.fasterxml.jackson.annotation.JsonFormat
+import com.fasterxml.jackson.module.kotlin.readValue
 import dateTimeFormat
 import typeid.TypeID
 import java.time.OffsetDateTime
 
 // TODO: linter gonna yell at us for this, but I want the typeid and serialization to be ez for now
 enum class MessageKind {
-    rfq, order, orderstatus
+  rfq, quote, close, order, orderstatus
 }
 
 class MessageMetadata(
-    val kind: MessageKind,
-    val to: String,
-    val from: String,
-    val id: TypeID,
-    val exchangeId: TypeID,
-    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = dateTimeFormat, timezone = "UTC")
-    val createdAt: OffsetDateTime
+  val kind: MessageKind,
+  val to: String,
+  val from: String,
+  val id: TypeID,
+  val exchangeId: TypeID,
+  @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = dateTimeFormat, timezone = "UTC")
+  val createdAt: OffsetDateTime
 )
 
-sealed interface MessageData
-
-abstract class Message<T : Any>(
-    val metadata: MessageMetadata,
-    val data: T,
-    var signature: String? = null
-) {
-    init {
-        when (metadata.kind) {
-            MessageKind.rfq -> require(data is RfqData)
-            MessageKind.order -> require(data is OrderData)
-            MessageKind.orderstatus -> require(data is OrderStatusData)
-        }
-
-        if (signature != null) {
-            verify()
-        } else {
-            validate()
-        }
-    }
 
-    private fun verify() {
-        validate()
+sealed class Message {
+  abstract val metadata: MessageMetadata
+  abstract val data: MessageData
+  abstract var signature: String?
 
-        // TODO detached payload sig check (regenerate payload and then check)
+  init {
+    if (signature != null) {
+      verify()
+    } else {
+      validate()
     }
+  }
+
+  fun verify() {
+    validate()
+
+    // TODO detached payload sig check (regenerate payload and then check)
+  }
 
-    private fun validate() {
-        // TODO validate against json schema
+  fun validate() {
+    // TODO validate against json schema
 //    val schema = schemaMap.get(metadata.kind.name)
 //    val jsonString = this.toString()
 //    schema.validateBasic(jsonString)
 //    if (output.errors != null) ...
-    }
+  }
 
-    // TODO - use web5 crypto and fix the types
-    fun sign(privateKey: String, kid: String) {
-        this.signature = "blah"
-    }
+  // TODO - use web5 crypto and fix the types
+  fun sign(privateKey: String, kid: String) {
+    this.signature = "blah"
+  }
 
-    override fun toString(): String {
-        return Mapper.writer().writeValueAsString(this)
-    }
+  fun toJsonString(): String {
```

Comment:
I think the pattern is just...

```suggestion
  fun toJson(): String {
```

Which implies it's a string. I think?

Reply:
which pattern?

Reply:
omg sorry i'm dumb yes this makes sense lmao